File: 2022/03/10/maxDepthBinaryTree.swift
// https://leetcode.com/problems/maximum-depth-of-binary-tree/
// Author: Andrew Jarombek
// Date: 3/10/2022

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init() { self.val = 0; self.left = nil; self.right = nil; }
    public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
    public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
        self.val = val
        self.left = left
        self.right = right
    }
}

func maxDepth(_ root: TreeNode?) -> Int {
    if (root == nil) {
        return 0
    }

    return 1 + max(maxDepth(root?.left), maxDepth(root?.right))
}

File: 2022/01/28/c/maxDepthBinaryTree.c
// Program: https://leetcode.com/problems/maximum-depth-of-binary-tree/
// Author: Andrew Jarombek
// Date: 1/28/2022

#include <stddef.h>
#include <assert.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int maxDepth(struct TreeNode* root) {
    if (root == NULL) {
        return 0;
    }

    int left = maxDepth(root->left) + 1;
    int right = maxDepth(root->right) + 1;

    return (left > right) ? left : right;
}

int main() {
    struct TreeNode node1 = {2, NULL, NULL};
    struct TreeNode node2 = {1, &node1, NULL};
    assert(maxDepth(&node2) == 2);
}
